packages/vite/src/node/config.ts

```diff
@@ -110,6 +110,10 @@ export interface UserConfig {
    * Or set to `false` to disable esbuild.
    */
   esbuild?: ESBuildOptions | false
+  /**
+   * Array of esbuild plugins to use.
+   */
+  esbuildPlugins?: EsbuildPlugin[]
```

Comment:
**suggestion:** IMO this should be part of `ESBuildOptions` and not its own property

Reply:
Yes indeed, however it may introduce breaking changes, as Vite config `esbuild` option (`ESBuildOptions`) are passed directly to esbuild **transform** function, not to **build** function. See: https://github.com/vitejs/vite/blob/0dc6e37298e2e3efda74a1042f37bd70e3f5d3a5/packages/vite/src/node/plugins/esbuild.ts#L52

We could think about:
1. Split the actual `esbuild` config into:
```js
esbuild: {
   transform: { /* the current `esbuild` options */ },
   build: { plugins: [/* the esbuild plugins */] },
},
```
but this is a *breaking* change

2. Keep the actual model and add `plugins` to it, or `build.plugins` to be clearer and allow for more flexibility in future developments:
```js
esbuild: {
   /* actual options */,
   build: { plugins: [/* the esbuild plugins */] },
},
```
and do `delete resolvedOptions.build` in `transformWithEsbuild` function before passing the options.

I am not convinced about the second solution.

I think the first one is much more clear, and we could go a step further by:

3. using the root of `ebuild` options (as today) for `include`, `exclude` and `jsxInject` options (which are **deleted** from `options` before the latter object is passed to `esbuild.transform` – so doing so we avoid this deletion and directly pass `options.transform` to `esbuild.transform()`, while `options.build` are passed directly to `esbuild.build()` with an appropriate merge):
```js
esbuild: {
   include: ...,
   exclude: ...,
   jsxInject: `import React from 'react'`,
   transform: {
     jsxFactory: 'h',
     jsxFragment: 'Fragment',
   },
   build: {
     plugins: [/* the esbuild plugins */],
     /* other future options */
   },
},
```

Reply:
🚀 I really like your proposal 👍 

@patak-js (@antfu) What are your thoughts about this?

Also IMO a breaking change shouldn't hold us back for such an improvement because it may be more future-proof

Reply:
Allright, if @patak-js and @antfu validate, I'll do the coding.
What are the roadmap for such a feature to be included in next minor ?
I'm pretty much impatient to use it ;-)

Reply:
I feel like since optimizeDeps is an entirely different build phase than the esbuild transform options, the settings should go into config.optimizeDeps. Here's a demonstration of that, perhaps this helps: https://github.com/vitejs/vite/pull/2991

I'd really like to get this functionality soon, since we're blocked by this in our conversion.